proposals/permissions-requestSiteAccess-api.md

```diff
@@ -0,0 +1,160 @@
+
+#### Other Alternatives considered
+
+- Action API is used to control the extension’s button in the browser’s toolbar. It’s exposed if the extension includes the "action" key in the manifest. This is troublesome since an extension could not have an action, but still want to show site access requests. We should not limit requests for extensions with actions
+- `permissions.addSiteAccessRequest()` resolves when request is accepted/rejected. An extension could be requesting site access and be granted site access through another mechanism (e.g changing site access in the extensions menu). We would either return a) true if we consider permissions granted through other mechanisms or b) false, because permission wasn't explicitely granted through the request. a) adds complexity and b) may cause confusion. Thus, we consider better to resolve whether the request is valid, and separately extension can listen whether permission is granted through `permissions.onAdded()`.
+
+### Manifest Changes
+
+None.
+
+### Custom Bindings
+
+None.
+
+### Design and Implementation
+
+#### Persistence
+
+Showing a site access request is never persisted.
+
+#### Alignment with Other Vendors and Action APIs
+
+Each browser can decide how they want to signal the extension site access requests. Browsers may decide to ignore the request (e.g due to noisiness, or if an extension has no visible UI).
+On Chrome’s side, we are exploring adding an ‘Allow’ chip in the extensions toolbar.
+
+## User Experience
+
+### UI Elements and User Visible-Effects
+
+Each browser will handle the request as desired. On Chrome’s side, we are considering showing an ‘Allow’ button in the toolbar for these requests. On click, it would grant persistent access to the site’s origin (which can be again withheld by the user)
+
+!["Site access request on Chrome toolbar"](./assets/permissions-requestSiteAccess-chrome-toolbar.png)
+
+We would also provide a way for users to silence the requests for an extension through a setting in the extensions menu and in chrome://extensions.
+
+!["Extensions menu on Chrome"](./assets/permissions-requestSiteAccess-chrome-menu.png)
+
+#### Attribution
+The site access request will be attributed directly to the extension.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+This API does not directly expose any sensitive data to the extension. However, it could lead to the extension gaining access to the site.
+
+### Abuse Scenarios
+
+Extensions can enable site access requests on every tab, which could be annoying to the user.
+
+#### Mitigations
+Attacks of annoyance are not in our threat model. Extensions can have annoying behaviors, but they risk uninstallation.
+On the browser side, we can place some restrictions to limit noisiness:
```

Comment:
The UI is typically bound to visible tabs. The document doesn't explicitly call out the desired behavior when the API is used on background tabs. What do you want to do there?

Supporting background tabs means that the browser needs to keep track of the access request for every tab and extension.

Limiting it to visible tabs only forces extensions to look at the current URL whenever the tab switches.

There are pros and cons to either approach. Given the lack of call-outs in this document, should I assume that you intend for the flag to stick independently of whether the tab is in the foreground?

Another option that solves this issue and limits abuse is to allow an extension to have only one pending request at any time, and clear the previous request when asked again. If justified we can also consider more than one request, but still a limited number.

Reply:
Good callout! I think we should allow developers to specify inactive tabs. Request will be shown by the browser when the user activated the tab. As you highlighted, it will require for the browser to keep track of the access request for every tab and extension. This shouldn't be problematic, it's a small amount of data.

Updated proposal (objectives section). What do you think?